README.md
# Open SAR Toolkit (OST)
[![Build Status](https://travis-ci.org/ESA-PhiLab/OpenSarToolkit.svg?branch=master)](https://travis-ci.org/ESA-PhiLab/OpenSarToolkit)

## Objective

This python package lowers the entry barrier for accessing and pre-processing 
Sentinel-1 data for land applications and allows users with little knowledge 
on SAR and python to produce various Analysis-Ready-Data products.

## Functionality

The Open SAR Toolkit (OST) bundles the full workflow for the generation of 
Analysis-Ready-Data (ARD) of Sentinel-1 for Land in a single high-level 
python package. It includes functions for data inventory and advanced sorting 
as well as downloading from various mirrors. The whole pre-processing is 
bundled in a single function and different types of ARD can be selected,
but also customised. OST does include advanced types of ARD such as combined
production of calibrated backscatter, interferometric coherence and the dual-
polarimetric H-A-Alpha decomposition. Time-series and multi-temporal statistics
(i.e. timescans) can be produced for each of these layers and the generation of 
spatially-seamless large-scale mosaic over time is possible a well.

The Open SAR Toolkit realises this by using an object-oriented approach, 
providing classes for single scene processing, GRD and SLC batch processing 
routines. The SAR processing itself relies on ESA's Sentinel-1 Toolbox as well 
as some geospatial python libraries and the Orfeo Toolbox for mosaicking.

You can find examplarotary Jupyter notebooks at 
https://github.com/ESA-PhiLab/OST_Notebooks for getting started. 

## Installation

### Docker 

**Update!**

Dockerhub is not permitting automatic builds. Therefore you need to
build your own docker image using the DOCKERFIlE https://raw.githubusercontent.com/ESA-PhiLab/OpenSarToolkit/main/Dockerfile

The resulting docker image contains the full package, 
including ESA's Sentinel-1 Toolbox, Orfeo Toolbox, Jupyter Lab as well
as the Open SAR Toolkit tutorial notebooks.

Docker installation is possible on various OS. Installation instructions can be 
found at https://docs.docker.com/install/

After docker is installed and running, launch the container with 
(adapt the path to the shared host folder and the name of the docke rimage at the very end):

```
docker run -it -p 8888:8888 -v /shared/folder/on/host:/home/ost/shared docker/image
```

The docker image automatically executes the jupyter lab and runs it on 
port 8888. You can find the address to the notebook on the command line where 
docker is running. Copy it into your favorites browser and replace 
127.0.0.1 with localhost.


### Manual installation

#### Dependencies

##### Sentinel Application Toolbox (SNAP)

OST bases mainly on the freely available SNAP toolbox for the 
SAR-specific processing routines. You can download SNAP from:

http://step.esa.int/main/download/

If you install SNAP into the standard directory, OST should have no problems 
to find the SNAP command line executable. Otherwise you need to define the path 
to the gpt file on your own during processing.

**Make sure to use SNAP 8 with the latest updates installed.**

##### Orfeo Toolbox

If you want to create mosaics between different swaths, OST will rely on the 
otbcli_Mosaic command from The Orfeo Toolbox. You can download Orfeo from:

https://www.orfeo-toolbox.org/download/

Make sure that the Orfeo bin folder is within your PATH variable to allow 
execution from command line.

##### Further dependencies (libs etc)

Ubuntu 18.04 and later:
```
sudo apt install python3-pip git libgdal-dev python3-gdal libspatialindex-dev nodejs npm libgfortran5
```

Any Operating system using (mini)conda https://www.anaconda.com/:

```
conda install pip gdal jupyter jupyterlab git matplotlib numpy rasterio imageio rtree geopandas fiona shapely matplotlib descartes tqdm scipy joblib retrying pytest pytest-cov nodejs
```

#### OST installation

You can use then use pip to install Open SAR Toolkit: 

```
pip install opensartoolkit
```


## Examples

### Ecuador VV-polarised Timescan Composite

Year: 2016

Sensor: Sentinel-1 C-Band SAR.

Acquisitions: 6 acquisitions per swath (4 swaths)

Output resolution: 30m

RGB composite:
  - Red: VV-maximum
  - Green: VV-minimum
  - Blue: VV-Standard deviation

![alt text](https://github.com/openforis/opensarkit/raw/master/shiny/www/ecuador_VV_max_min_std.png)

### Ethiopia VV-VH polarised Timescan Composite

Year: 2016-2017

Sensor: Sentinel-1 C-Band SAR.

Acquisitions: 7 acquisitions per swath (about 400 scenes over 8 swaths)

Output resolution: 30m

RGB composite:
  - Red: VV-minimum
  - Green: VH-minimum
  - Blue: VV-Standard deviation

![alt text](https://github.com/openforis/opensarkit/raw/master/shiny/www/eth_vvvh_ts.jpeg)


## A note on its origin

Open SAR Toolkit was initially developed at the Food and Agriculture 
Organization of the United Nations under the SEPAL project 
(https://github.com/openforis/sepal) between 2016-2018. 
It is still available there (https://github.com/openforis/opensarkit), 
but has been completely re-factored and transferred into a simpler and 
less-dependency rich python3 version, which can be found on this page here. 
Instead of using R-Shiny as a GUI, the main interface are now Jupyter notebooks 
that are developed in parallel to this core package and should help to get started.
(https://github.com/ESA-PhiLab/OST_Notebooks) 


## Author

* Andreas Vollrath, FAO (ex-ESA philab)
* Petr Sevcik, EOX
* James Wheeler, ESA philab


CHANGELOG.md

## 0.12.15 (2021-12-28)

### Refactor

- compress big files

## 0.12.14 (2021-12-17)

### Fix

- document contributing

## 0.12.13 (2021-12-17)

### Fix

- typo in yaml format
- install gdal before publishing

## 0.12.12 (2021-12-17)

### Fix

- the if need to be set after job name

## 0.12.11 (2021-12-17)

### Fix

- only push to pipy if the commit message is a bumping message

## 0.12.10 (2021-12-17)

### Fix

- no v in tag names
- branch name in github actions
- create a fix to trigger the bump

### Refactor

- drop useless with_suffix
- drop usage of joinpath
- lint
- compliant with PEP and Black

## 0.12.5 (2021-09-24)

## 0.10.1 (2020-08-27)


s1/metadata.py
__author__ = "Andreas Vollrath"

import sys
import glob
import getpass
import urllib
import xml.dom.minidom
import numpy as np
import geopandas as gpd
import xml.etree.ElementTree as ET

from shapely.wkt import loads
from urllib.error import URLError

from ost.s1 import search


class s1Metadata:
    """
    Get metadata from Sentinel-1 scene id
    """

    num_of_scenes = 0

    def __init__(self, scene_id):
        self.scene_id = scene_id
        self.mission_id = scene_id[0:3]
        self.mode_beam = scene_id[4:6]
        self.product_type = scene_id[7:10]
        self.res_class = scene_id[10]
        self.proc_level = scene_id[12]
        self.pol_mode = scene_id[14:16]
        self.start_date = scene_id[17:25]
        self.start_time = scene_id[26:32]
        self.stop_date = scene_id[33:41]
        self.stop_time = scene_id[42:48]
        self.abs_orbit = scene_id[49:55]
        self.data_take_id = scene_id[57:62]
        self.unique_id = scene_id[63:]
        self.year = scene_id[17:21]
        self.month = scene_id[21:23]
        self.day = scene_id[23:25]

        # Calculate the relative orbit out of absolute orbit
        # (from Peter Meadows (ESA) @
        # http://forum.step.esa.int/t/sentinel-1-relative-orbit-from-filename/7042)
        if self.mission_id == 'S1A':
            self.orbit_offset = 73
            self.satellite = "Sentinel-1A"
        elif self.mission_id == 'S1B':
            self.orbit_offset = 27
            self.satellite = "Sentinel-1A"

        self.rel_orbit = (((int(self.abs_orbit)
                            - int(self.orbit_offset)) % 175) + 1)

        # get acquisition mode
        if self.mode_beam == 'IW':
            self.acq_mode = "Interferometric Wide Swath"
        elif self.mode_beam == 'SM':
            self.acq_mode = "Stripmap"
        elif self.mode_beam == 'EW':
            self.acq_mode = "Extra-Wide swath"
        elif self.mode_beam == 'WV':
            self.acq_mode = "Wave"

        # get acquisition mode
        if self.product_type == 'GRD':
            self.p_type = "Ground Range Detected (GRD)"
        elif self.product_type == 'SLC':
            self.p_type = "Single-Look Complex (SLC)"
        elif self.product_type == 'OCN':
            self.p_type = "Ocean"
        elif self.product_type == 'RAW':
            self.p_type = "Raw Data (RAW)"

        # increment class variable for every scene read
        s1Metadata.num_of_scenes += 1

    def s1Info(self):

        print(" -------------------------------------------------")
        print(" Scene Information:")
        print(" Scene Identifier:        " + str(self.scene_id))
        print(" Satellite:               " + str(self.satellite))
        print(" Acquisition Mode:        " + str(self.acq_mode))
        print(" Processing Level:        " + str(self.proc_level))
        print(" Product Type:            " + str(self.p_type))
        print(" Acquisition Date:        " + str(self.start_date))
        print(" Start Time:              " + str(self.start_time))
        print(" Stop Time:               " + str(self.stop_time))
        print(" Absolute Orbit:          " + str(self.abs_orbit))
        print(" Relative Orbit:          " + str(self.rel_orbit))
        print(" -------------------------------------------------")

    def s1IPTpath(self, basePath='/eodata/Sentinel-1'):

        path = '{}/SAR/{}/{}/{}/{}/{}.SAFE'.format(basePath, self.product_type,
                                                   self.year, self.month,
                                                   self.day, self.scene_id)

        return path
    
    
    def s1DwnPath(self, dwnDir):

        path = '{}/SAR/{}/{}/{}/{}/{}.zip'.format(dwnDir, self.product_type,
                                                   self.year, self.month,
                                                   self.day, self.scene_id)

        return path

    def s1EsaUuidFromId(self, opener):

        # construct the basic the url
        baseURL = ('https://scihub.copernicus.eu/apihub/odata/v1/'
                   'Products?$filter=')
        action = urllib.request.quote('Name eq \'{}\''.format(self.scene_id))
        # construct the download url
        url = baseURL + action

        try:
            # get the request
            req = opener.open(url)
        except URLError as e:
            if hasattr(e, 'reason'):
                print(' We failed to connect to the server.')
                print(' Reason: ', e.reason)
                sys.exit()
            elif hasattr(e, 'code'):
                print(' The server couldn\'t fulfill the request.')
                print(' Error code: ', e.code)
                sys.exit()
        else:
            # write the request to to the response variable
            # (i.e. the xml coming back from scihub)
            response = req.read().decode('utf-8')

            # parse the xml page from the response
            dom = xml.dom.minidom.parseString(response)

            # loop thorugh each entry (with all metadata)
            for node in dom.getElementsByTagName('entry'):
                downloadURL = node.getElementsByTagName(
                                            'id')[0].firstChild.nodeValue
                uuid = downloadURL.split('(\'')[1].split('\')')[0]

        return uuid

    def s1EsaSceneUrl(self, opener):

        uuid = self.s1EsaUuidFromId(opener)
        # scihub url
        scihubURL = 'https://scihub.copernicus.eu/apihub/odata/v1/Products'
        # construct the download url
        urlDownload = '{}(\'{}\')/$value'.format(scihubURL, uuid)

        return urlDownload

    def s1EsaMd5Url(self, opener):

        uuid = self.s1EsaUuidFromId(opener)
        scihubURL = 'https://scihub.copernicus.eu/apihub/odata/v1/Products'
        urlDownload = '{}(\'{}\')/Checksum/Value/$value'.format(scihubURL,
                                                                uuid)
        return urlDownload

    def s1EsaAnnoUrl(self, opener):

        uuid = self.s1EsaUuidFromId(opener)

        print(' INFO: Getting URLS of annotation files'
              ' for S1 product: {}.'.format(self.scene_id))
        scihubURL = 'https://scihub.copernicus.eu/apihub/odata/v1/Products'
        annoPath1 = ('(\'{}\')/Nodes(\'{}.SAFE\')/Nodes(\'annotation\')/'
                     'Nodes'.format(uuid, self.scene_id))
        url = scihubURL + annoPath1

        try:
            # get the request
            req = opener.open(url)
        except URLError as e:
            if hasattr(e, 'reason'):
                print(' We failed to connect to the server.')
                print(' Reason: ', e.reason)
                sys.exit()
            elif hasattr(e, 'code'):
                print(' The server couldn\'t fulfill the request.')
                print(' Error code: ', e.code)
                sys.exit()
        else:
            # write the request to to the response variable
            # (i.e. the xml coming back from scihub)
            response = req.read().decode('utf-8')

            # parse the xml page from the response
            dom = xml.dom.minidom.parseString(response)

            urlList = []
            # loop thorugh each entry (with all metadata)
            for node in dom.getElementsByTagName('entry'):
                downloadURL = node.getElementsByTagName(
                                                'id')[0].firstChild.nodeValue

                if downloadURL[-6:-2] == '.xml':
                    urlList.append('{}/$value'.format(downloadURL))

        return urlList

    def s1IPTAnno(self):

        colNames = ['SceneID', 'Date', 'SwathID', 'BurstID',
                    'BurstNr', 'geometry']
        gdfFull = gpd.GeoDataFrame(columns=colNames)

        for annoFile in glob.glob(
                '{}/annotation/*xml'.format(self.s1IPTpath())):
            # parse the xml page from the response
            gdf = self.s1BurstInfo(ET.parse(annoFile))

            gdfFull = gdfFull.append(gdf)

        return gdfFull.drop_duplicates(['BurstID'], keep='first')

    def s1EsaAnno(self, uname=None, pword=None):

        colNames = ['SceneID', 'Date', 'SwathID', 'BurstID',
                    'BurstNr', 'geometry']
        gdfFull = gpd.GeoDataFrame(columns=colNames)

        baseURL = "https://scihub.copernicus.eu/apihub/"
        # ask for username and password in case you have
        # not defined as command line options

        if uname is None:
            print(' If you do not have a Copernicus Scihub user'
                  ' account go to: https://scihub.copernicus.eu')
            uname = input(' Your Copernicus Scihub Username:')
        if pword is None:
            pword = getpass.getpass(' Your Copernicus Scihub Password:')

        # get connected to scihub
        opener = search.scihubConnect(baseURL, uname, pword)

        annoList = self.s1EsaAnnoUrl(opener)

        for url in annoList:
            try:
                # get the request
                req = opener.open(url)
            except URLError as e:
                if hasattr(e, 'reason'):
                    print(' We failed to connect to the server.')
                    print(' Reason: ', e.reason)
                    sys.exit()
                elif hasattr(e, 'code'):
                    print(' The server couldn\'t fulfill the request.')
                    print(' Error code: ', e.code)
                    sys.exit()
            else:
                # write the request to to the response variable
                # (i.e. the xml coming back from scihub)
                response = req.read().decode('utf-8')

                ETroot = ET.fromstring(response)

                # parse the xml page from the response
                gdf = self.s1BurstInfo(ETroot)

                gdfFull = gdfFull.append(gdf)

        return gdfFull.drop_duplicates(['BurstID'], keep='first')

    def s1BurstInfo(self, ETroot):
        '''
        This functions expects an xml string from a Sentinel-1 SLC
        annotation file and extracts relevant information for burst
        identification as a GeoPandas GeoDataFrame.

        Much of the code is taken from RapidSAR
        package (once upon a time on github).
        '''

        colNames = ['SceneID', 'Date', 'SwathID',
                    'BurstID', 'BurstNr', 'geometry']
        gdf = gpd.GeoDataFrame(columns=colNames)

        track = self.rel_orbit
        acqDate = self.start_date

        # pol = root.find('adsHeader').find('polarisation').text
        swath = ETroot.find('adsHeader').find('swath').text
        linesPerBurst = np.int(ETroot.find('swathTiming').find(
                                                    'linesPerBurst').text)
        pixelsPerBurst = np.int(ETroot.find('swathTiming').find(
                                                    'samplesPerBurst').text)
        burstlist = ETroot.find('swathTiming').find('burstList')
        geolocGrid = ETroot.find('geolocationGrid')[0]
        first = {}
        last = {}

        # Get burst corner geolocation info
        for geoPoint in geolocGrid:
            if geoPoint.find('pixel').text == '0':
                first[geoPoint.find('line').text] = np.float32(
                        [geoPoint.find('latitude').text,
                         geoPoint.find('longitude').text])
            elif geoPoint.find('pixel').text == str(pixelsPerBurst-1):
                last[geoPoint.find('line').text] = np.float32(
                        [geoPoint.find('latitude').text,
                         geoPoint.find('longitude').text])

        for i, b in enumerate(burstlist):
            firstline = str(i*linesPerBurst)
            lastline = str((i+1)*linesPerBurst)
            aziAnxTime = np.float32(b.find('azimuthAnxTime').text)
            burstid = np.int32(np.round(aziAnxTime*10))
            burstid = 'T{}_{}_{}'.format(track, swath, burstid)
            # first and lastline sometimes shifts by 1 for some reason?
            try:
                firstthis = first[firstline]
            except:
                firstline = str(int(firstline)-1)
                try:
                    firstthis = first[firstline]
                except:
                    print('First line not found in annotation file')
                    firstthis = []
            try:
                lastthis = last[lastline]
            except:
                lastline = str(int(lastline)-1)
                try:
                    lastthis = last[lastline]
                except:
                    print('Last line not found in annotation file')
                    lastthis = []
            corners = np.zeros([4, 2], dtype=np.float32)

            # Had missing info for 1 burst in a file, hence the check
            if len(firstthis) > 0 and len(lastthis) > 0:
                corners[0] = first[firstline]
                corners[1] = last[firstline]
                corners[3] = first[lastline]
                corners[2] = last[lastline]

            wkt = 'POLYGON (({} {},{} {},{} {},{} {},{} {}))'.format(
                        np.around(float(corners[0, 1]), 3),
                        np.around(float(corners[0, 0]), 3),
                        np.around(float(corners[3, 1]), 3),
                        np.around(float(corners[3, 0]), 3),
                        np.around(float(corners[2, 1]), 3),
                        np.around(float(corners[2, 0]), 3),
                        np.around(float(corners[1, 1]), 3),
                        np.around(float(corners[1, 0]), 3),
                        np.around(float(corners[0, 1]), 3),
                        np.around(float(corners[0, 0]), 3))

            gDict = {'SceneID': self.scene_id, 'Date': acqDate,
                     'SwathID': swath, 'BurstID': burstid,
                     'BurstNr': i+1, 'geometry': loads(wkt)}
            gdf = gdf.append(gDict, ignore_index=True)

        return gdf

    def s1ASFURL(self):

        asfURL = 'https://datapool.asf.alaska.edu'

        if self.mission_id == 'S1A':
            mission = 'SA'
        elif self.mission_id == 'S1B':
            mission = 'SB'

        if self.product_type == 'SLC':
            pType = self.product_type
        elif self.product_type == 'GRD':
            pType = 'GRD_{}{}'.format(self.res_class, self.pol_mode[0])

        productURL = '{}/{}/{}/{}.zip'.format(asfURL, pType,
                                              mission, self.scene_id)
        return productURL
